File: homework/src/main/java/com/mikheev/homework/repositories/impl/BookCustomRepositoryImpl.java
package com.mikheev.homework.repositories.impl;

import com.mikheev.homework.domain.Book;
import com.mikheev.homework.repositories.BookCustomRepository;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityGraph;
import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.Map;
import java.util.Optional;

@Repository
public class BookCustomRepositoryImpl implements BookCustomRepository {

    @PersistenceContext
    private EntityManager em;

    @Override
    public Optional<Book> findById(long id, String entityGraphName) {
        EntityGraph entityGraph = em.getEntityGraph(entityGraphName);
        return Optional.ofNullable(em.find(Book.class, id, Map.of("javax.persistence.fetchgraph", entityGraph)));
    }
}

File: homework/src/main/kotlin/com/mikheev/homework/repositories/impl/BookCustomRepositoryImpl.kt
package com.mikheev.homework.repositories.impl

import com.mikheev.homework.domain.Book
import com.mikheev.homework.repositories.BookCustomRepository
import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext
import org.springframework.stereotype.Repository
import java.util.*

@Repository
class BookCustomRepositoryImpl : BookCustomRepository {

    @PersistenceContext
    lateinit var em: EntityManager

    override fun findById(id: Long, entityGraphName: String): Optional<Book> {
        val entityGraph = em.getEntityGraph(entityGraphName)
        return Optional.ofNullable(em.find(Book::class.java, id, mapOf("jakarta.persistence.fetchgraph" to entityGraph)))
    }
}